### PR TITLE
Trying to get some OpInfo based test for MPS

### PR DESCRIPTION
**Pull request by albanD**
This does a few changes:
- new `TORCH_CHECK` to make sure the mps backend returns nice error instead of hard crash
- make print behavior smoother by converting to cpu 
- Add an OpInfo-based test to make sure mps output matches the cpu one

There are several issues with the opinfo test:
- One test just hangs forever
- Some function still hard crash and probably need more TORCH_CHECK
- Many functions are failing in a flaky way: they sometimes pass, sometimes fail (usually because the output is all 0s)

Next steps:
- Fix all the things in the `BLOCKLIST`
- Make `ALLOWLIST_OP` not depend on dtype anymore (so that we can just enable ops directly via a boolean flag). This is not done now because many op fail int/bool dtypes.
- Move `ALLOWLIST_OP` to be a boolean flag directly in `common_method_invocations.py` like `supports_mps=True`.
- Make `supports_mps=True` the default for all ops
- Extend this to check that gradients match

What needs to happen before merging:
- Ensure all the `TORCH_CHECK()` are not overly restrictive
- Ensure the newly added test is not flaky (add more stuff to the `BLOCKLIST`)

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 1]
```diff
@@ -200,6 +202,10 @@ Tensor mps_convolution_backward_input(
             weight{ weight_t, "weight", 2 };
   checkAllSameType(c, {grad_output, weight});
   checkAllSameGPU(c, {grad_output, weight});
+  TORCH_CHECK(isFloatingType(weight_t.dtype().toScalarType()), "Only floating point dtypes are supported for conv_transpose on MPS.");
```
[the remark this answers is not in the document]
Comment (albanD) [reply aten/src/ATen/native/mps/operations/Convolution.mm:205]: Ho good point!
mps tensors are plain Tensors (just like cuda ones). So yes that method should be there.

File: test/test_mps.py [hunk 2]
```diff
@@ -4288,7 +4291,383 @@ def test_legacy_constructor(self):
...
+        'nonzero': None,
+        'bool': None,
+    }
+
+    # Used for accept mode only
+    NEW_ALLOW_LIST = defaultdict(list)
+
+    @ops(op_db)
+    def test_output_match(self, device, dtype, op):
+        self.assertEqual(device, "cpu")
```
Comment (mruberry) [inline test/test_mps.py:4607]: This test might make more sense to model as an mps device test that is only instantiated when mps is around (with something like an `@onlyMPS` decorator) since we assume the CPU is always an available device for testing

File: test/test_mps.py [hunk 3]
```diff
@@ -4288,7 +4291,383 @@ def test_legacy_constructor(self):
...
+        if not torch.backends.mps.is_available():
+            self.skipTest("MPS is not available")
+
+        if dtype in [torch.bfloat16, torch.double, torch.int8]:
```
Comment (mruberry) [inline test/test_mps.py:4611]: The `@op`s decorator can be extended with an `allowed_dtypes` kwarg that will let you filter the dtype instantiations as you like

File: test/test_mps.py [hunk 4]
```diff
@@ -4288,7 +4291,383 @@ def test_legacy_constructor(self):
...
+        try:
+            cpu_samples = op.sample_inputs(device, dtype)
+
+            for cpu_sample in cpu_samples:
+                cpu_args = [cpu_sample.input] + list(cpu_sample.args)
+                cpu_kwargs = cpu_sample.kwargs
+                mps_args = tuple(arg.to("mps") if isinstance(arg, torch.Tensor) else arg for arg in cpu_args)
+                mps_kwargs = {k: v.to("mps") if isinstance(v, torch.Tensor) else v for k, v in cpu_kwargs.items()}
+
+                if True or op.supports_mps:
```
Comment (mruberry) [inline test/test_mps.py:4646]: `True or`... ?

File: test/test_mps.py [hunk 5]
```diff
@@ -4288,7 +4291,383 @@ def test_legacy_constructor(self):
...
+        key = op.name + op.variant_test_name
+        if key in self.BLOCKLIST:
+            if self.BLOCKLIST[key] is None or dtype in self.BLOCKLIST[key]:
+                self.skipTest(f"Running test with {op.name} hangs so skipping")
+
+        # Make this an expecttest manually
+        # When this env variable is set, generate a new ALLOWLIST_OP
+        # that reflects the current state of what passes or not
+        if os.environ.get("EXPECTTEST_ACCEPT", None) == "1":
+            generate_new_truth = True
```
Comment (mruberry) [inline test/test_mps.py:4626]: quite the variable name

File: test/test_mps.py [hunk 6]
```diff
@@ -4288,7 +4291,383 @@ def test_legacy_constructor(self):
...
+                    with self.assertRaisesRegex(NotImplementedError, "not impls?"):
+                        mps_out = op(*cpu_args, **mps_kwargs)
+        except Exception as e:
+            if not generate_new_truth:
+                raise e
+        else:
+            if generate_new_truth:
+                self.NEW_ALLOW_LIST[op.name].append(str(dtype))
+
+        # We could write it only once. But I don't know how to detect that the current test is the last one
```
Comment (mruberry) [inline test/test_mps.py:4663]: Holy cow though -- can't this at least condition on whether the list was changed or not by the above?

**Review by mruberry — APPROVED** [review]
Approving OpInfo parts for velocity but see inline comments for some suggestions

File: test/test_mps.py (hunk 2 above)
Comment (albanD) [reply test/test_mps.py:4607]: My understanding is that to be able to do that, I would have to add MPS as a new backend for the device-generic test system? Since we don't have that right now, that's why I went with CPU. Is that inacurate?

File: test/test_mps.py [hunk 7]
```diff
@@ -4288,7 +4291,383 @@ def test_legacy_constructor(self):
...
+                if str(dtype) not in self.ALLOWLIST_OP[op.name]:
+                    self.skipTest(f"{op.name} is in the allow list for MPS but {dtype} is excluded")
+
+        try:
+            cpu_samples = op.sample_inputs(device, dtype)
+
+            for cpu_sample in cpu_samples:
+                cpu_args = [cpu_sample.input] + list(cpu_sample.args)
+                cpu_kwargs = cpu_sample.kwargs
+                mps_args = tuple(arg.to("mps") if isinstance(arg, torch.Tensor) else arg for arg in cpu_args)
```
[the remark this answers is not in the document]
Comment (albanD) [reply test/test_mps.py:4643]: I didn't knew it existed!

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -19912,10 +19912,12 @@ def __init__(
 sparse_masked_reduction_ops = [op for op in sparse_reduction_ops if op.name.startswith('_masked.')]
 
 # TODO: review porting these to make_tensor
-def index_variable(shape, max_indices, device=torch.device('cpu')):
+def index_variable(shape, max_indices, device='cpu'):
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:19915]: Yes, this was a surprising one to find.

File: test/test_mps.py (hunk 6 above)
Comment (albanD) [reply test/test_mps.py:4663]: Sure.
Note that this is only ever run locally. And this whole code should go away in a month once we can just integrate with OpInfo properly.

File: aten/src/ATen/native/mps/operations/Repeat.mm [hunk 9]
```diff
@@ -75,6 +75,11 @@ Tensor repeat_mps(const Tensor& self, IntArrayRef repeats) {
   TORCH_CHECK(repeats.size() >= (size_t)self.dim(),
            "Number of dimensions of repeat dims can not be smaller than number of dimensions of tensor");
 
+  if (self.sizes() == repeats) {
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Repeat.mm:78]: This failing a test in test_mps:

```
    def test_repeat(self):
        def helper(shape, repeats):

            cpu_x = torch.randn(shape, device='cpu', dtype=torch.float, requires_grad=True)
            x = cpu_x.detach().clone().to('mps').requires_grad_()

            y = x.repeat(repeats)
            ref_y = cpu_x.repeat(repeats)

            cpu_grad = torch.randn(ref_y.shape)
            grad = cpu_grad.to('mps')

            y.backward(gradient=grad)
            ref_y.backward(gradient=cpu_grad)

            self.assertEqual(y, ref_y)
            self.assertEqual(x.grad, cpu_x.grad)

        helper((2, 3, 4, 5), (2, 3, 4, 5))
        helper((2, 3, 4), (4, 3, 2, 5, 7, 2))
        helper((3, 4, 5), (2, 3, 4, 5))
        helper((3, 4, 5), (2, 2, 2))
```

File: aten/src/ATen/native/mps/operations/Linear.mm [hunk 10]
```diff
@@ -25,6 +25,12 @@ Tensor _mps_linear(
 
   using namespace mps;
 
+  TORCH_CHECK(input.dtype() == weight.dtype(), "tensors must have the same dtype");
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Linear.mm:28]: Some of the test_mps.py tests:
```
    def test_linear2D_no_bias(self):
        self._linear_helper(in_features=2, out_features=3, shape=((4, 2)), bias=False, backward_pass=False)
```
are failing, here the bias can have UNKNOWN_SCALAR type or none dtype

```
test_linear2D_no_bias_backward (__main__.TestMPS) ... 
Input float
weight float
bias UNKNOWN_SCALAR
ERROR
```

Comment (kulinseth) [reply aten/src/ATen/native/mps/operations/Linear.mm:28]: I think we should relax the bias type check.

File: aten/src/ATen/native/mps/operations/Repeat.mm (hunk 9 above)
Comment (kulinseth) [reply aten/src/ATen/native/mps/operations/Repeat.mm:78]: We should have something like:
```

+  if (num_repeat_dims == 0) {
+    return at::native::empty_mps(
+                        {}, self.scalar_type(),
+                        c10::nullopt,
+                        kMPS,
+                        c10::nullopt,
+                        c10::nullopt);
+  }
```

**Review by kulinseth — COMMENTED** [review]
Can you please take a look at those ?

File: aten/src/ATen/native/mps/operations/Repeat.mm (hunk 9 above)
Comment (albanD) [reply aten/src/ATen/native/mps/operations/Repeat.mm:78]: Yes, that check is just not working at all. Fixed.

File: aten/src/ATen/native/mps/operations/Linear.mm (hunk 10 above)
Comment (albanD) [reply aten/src/ATen/native/mps/operations/Linear.mm:28]: Ho the issue is that the bias can be present but undefined. Adding support for that.

**Comment by albanD** [comment]
Closing as this landed in another PR.
